ENH: provide standard BaseIndexers in pandas.api.indexers

AlexKirko opened the pull request:
- [X] closes #33201
- [X] 1 test modified / 1 passed
- [X] passes `black pandas`
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] whatsnew entry

This PR adds a forward-looking `pandas.api.indexers.BaseIndexer` subclass implementation and exposes it to users as discussed in #33180 .

### Exposing other indexers

It is also possible to expose a backward-looking indexer simply by exposing `pandas.core.window.indexers.FixedWindowIndexer` through the API. Seems redundant to me as this is basically the same as supplying an integer as the window argument when creating a rolling object but it can be done for consistency. I'd be grateful for a discussion.

### Note on tests

The class is incorporated into `test_rolling_forward_window` in `test_base_indexer.py` and this serves as a test that the new indexer is working correctly.

### Notes on performance
I've tested the implementation and it shows comparable performance to `pandas.core.window.indexers.FixedWindowIndexer`. In my tests the new indexer was slightly faster, either by luck, or because I streamlined the backward-looking indexer a bit when making the forward-looking one.

AlexKirko committed:
ENH: initial indexer implementation

AlexKirko committed:
FIX: fix brackets

AlexKirko committed:
ENH: expose FixedForwardWIndowIndexer through API

AlexKirko committed:
DOC: add whatsnew entry

AlexKirko committed:
CLN: run black

AlexKirko committed:
DOC: add example to Rolling docstring

WillAyd commented:
/azp run

jreback commented:
looks good, if you can rebase and make the change to use the FixedForward indexer. Also I think worthile adding a benchmark.

jreback commented:
@TomAugspurger @jorisvandenbossche @jbrockmendel if any comments.

AlexKirko committed:
TST: use new class in test_rolling_forward_window

AlexKirko committed:
TST: add benchmark

AlexKirko committed:
fix benchmark

AlexKirko commented:
Travis failing has nothing to do with this PR. See #33280 
Waiting for merge.

AlexKirko committed:
Merge branch 'master' into forward-indexer

mroeschke commented:
Could you document this new indexer around here: https://pandas.pydata.org/pandas-docs/stable/user_guide/computation.html#custom-window-rolling

AlexKirko committed:
DOC: document the indexer in the user guide

AlexKirko committed:
DOC: edit the text in the user_guide

AlexKirko commented:
Found a bug, needs a bit more work.

AlexKirko committed:
fix the benchmark

AlexKirko commented:
@mroeschke Added documentation to the custom window rolling subsection. Can introduce a new subsection, but I think it fits in. Please tell me what you think.

AlexKirko commented:
@jreback 
Rebased, changed the test, added a benchmark to the suite. All green.

jbrockmendel commented:
LGTM

AlexKirko committed:
DOC: add versionadded to user_guide

AlexKirko commented:
Small change: had forgotten to include versionadded in the user guide. Included it.

jreback commented on doc/source/user_guide/computation.rst:
```diff
@@ -571,6 +571,29 @@ and we want to use an expanding window where ``use_expanding`` is ``True`` other
+For some problems knowledge of the future is available for analysis. For example, this occurs when
+each data point is a full time series read from an experiment, and the task is to extract underlying
+conditions. In these cases it can be useful to perform forward-looking rolling window computations.
+``FixedForwardWindowIndexer`` class is available for this purpose. This ``BaseIndexer`` subclass
```
you can use the ref here, e.g. the one you are using in the whatsnew

AlexKirko replied:
@jreback I've included a ref in whatsnew, pointing to this subsection of the user guide. If you meant something different, please tell me.

jreback replied:
no I mean a ref on FixedForwardIndexer here that points to the api (makes a clickable link), and one for BaseIndexer

jreback commented on doc/source/user_guide/computation.rst:
```diff
@@ -571,6 +571,29 @@ and we want to use an expanding window where ``use_expanding`` is ``True`` other
+``FixedForwardWindowIndexer`` class is available for this purpose. This ``BaseIndexer`` subclass
+implements a closed fixed-width forward-looking rolling window, and we can use it as follows:
+
+.. code-block:: ipython
```
use an ipython block here

AlexKirko replied:
@jreback 
Sorry, could you clarify? To me, my ipython block looks right, both in raw and in the docs I compiled locally. Do I need to do something differently?

jreback replied:
we don't use code-blocks on executable code. pls use an ipython block so that the doc-build executes this, that way it will never get out of date.

jreback commented on pandas/core/window/indexers.py:
```diff
@@ -120,3 +120,23 @@ def get_window_bounds(
+    """Calculate fixed forward-looking rolling window bounds,"""
+
+    @Appender(get_window_bounds_doc)
+    def get_window_bounds(
```
is this a sufficient doc-string?

AlexKirko replied:
Expanded and clarified the class docstring a bit and added examples directly to it. I think now it's clear just what the new class does.
The `get_window_bounds` docstring is the same for every `BaseIndexer` subclass. We could add a particular docstring to the function instead of using the Appender, but I don't think it would add much clarity for the user.

jreback commented on pandas/core/window/indexers.py:
```diff
@@ -120,3 +120,23 @@ def get_window_bounds(
+
+
+class FixedForwardWindowIndexer(BaseIndexer):
+    """Calculate fixed forward-looking rolling window bounds,"""
```
can you add Examples section here

AlexKirko replied:
Done.

jreback commented on pandas/core/window/indexers.py:
```diff
@@ -120,3 +120,23 @@ def get_window_bounds(
+    def get_window_bounds(
+        self,
+        num_values: int = 0,
+        min_periods: Optional[int] = None,
```
this should raise if min_periods, center, closed are not None

AlexKirko replied:
Implemented it a bit differently, pinged you in the comment on the implementation.

AlexKirko committed:
raise if min_periods, center, or closed is not None

AlexKirko committed:
DOC: expand the class docstring

AlexKirko committed:
DOC: add examples to the class docstring

AlexKirko committed:
DOC: add ref from whatsnew to the user guide

AlexKirko committed:
rollback raising behavior

AlexKirko committed:
add errors for center and closed

AlexKirko commented:
Looks like the doc build fail is being handled in #33309

AlexKirko committed:
CLN: run black

AlexKirko committed:
CLN: remove unnecessary quotes

AlexKirko commented on pandas/core/window/indexers.py:
```diff
@@ -120,3 +120,53 @@ def get_window_bounds(
+        if closed is not None:
+            raise ValueError(
+                "Forward-looking windows don't support setting the closed argument"
+            )
```
@jreback
Implemented the error-raising behavior this way. Reasons below.
As far as I've been able to test, the arguments are being set by the `rolling` object in a general way and outside the control of the user. They are passed whether a particular `BaseIndexer` implementation needs them or not.
`min_periods` is getting set to `rolling.min_periods`. Same stuff for `center`: in my tests it got set to `False` by default. Only `closed` got passed as `None`.
I wrote the error-raising behavior with this in mind. We should raise if something that doesn't make sense got passed, but testing for None is too crude for our purposes.

AlexKirko committed:
Merge branch 'master' into forward-indexer

AlexKirko commented:
@jreback 
All green, pinged you on relevant changes. I think I've addressed everything, but I'm not sure if I understood you correctly on adding refs and the ipython block. Please tell me if I misunderstood anything.

jreback commented on pandas/core/window/indexers.py:
```diff
@@ -120,3 +120,53 @@ def get_window_bounds(
+        if center:
+            raise ValueError("Forward-looking windows can't have center=True")
+        if closed is not None:
+            raise ValueError(
```
need a test that hits these

AlexKirko replied:
Done. Added to the test.

AlexKirko committed:
DOC: replace code-block with ipython block

AlexKirko committed:
DOC: expose docstring, make links clickable

AlexKirko committed:
TST: add error raising tests

AlexKirko commented:
@jreback
Thanks for the clarification! Sorry for getting confused: first time adding a public class.
1. Added an entry to the `window.rst` so that the docstrings get added to the API reference output.
2. Made the links in the user guide clickable and tested locally.
3. Replaced the `code-block` with an `ipython` block. Thanks.
4. Added `pytest.raises` to the test to test error-raising behavior. I think having them there works, but please tell me if you'd rather have them in a separate test.

AlexKirko committed:
Merge branch 'master' into forward-indexer

AlexKirko committed:
CLN: clean up formatting in docs

AlexKirko committed:
CLN: remove unnecessary BaseIndexer import

AlexKirko committed:
CLN: remove unnecessary new line in computation.rst

mroeschke commented on doc/source/reference/window.rst:
```diff
@@ -85,3 +85,14 @@ Base class for defining custom window boundaries.
 
    api.indexers.BaseIndexer
+
+Forward-looking fixed window indexer
```
I don't think you need to add a whole new section for this. Just place `api.indexers.FixedForwardWindowIndexer` right below `api.indexers.BaseIndexer`

AlexKirko replied:
That looks better, thanks. Still learning how our docs are built, sorry.

AlexKirko committed:
DOC: move new indexer in window reference, delete extra

jreback commented:
thanks @AlexKirko very nice!
